File: CPP/ByRound/1881/1881_G_Anya_and_the_Mysterious_String.cpp
// Problem: CF 1881 G - Anya and the Mysterious String
// https://codeforces.com/contest/1881/problem/G

/*
Problem: Determine if a substring of a modified string is "beautiful"
         (i.e., contains no palindromic substrings of length >= 2).

Approach:
- Use Binary Indexed Trees (Fenwick Trees) to efficiently handle range updates and queries.
- Maintain two BITs:
  - bit1: tracks differences between adjacent characters (d[i] = s[i] - s[i-1])
  - bit2: tracks differences between characters two positions apart (e[i] = s[i] - s[i-2])
- A substring is beautiful if:
  - No adjacent characters are the same (bit1 query)
  - No three-character palindromes exist (bit2 query)
- For range updates, apply delta to specific indices using BIT range update technique.

Time Complexity: O((n + m) * log n) per test case
Space Complexity: O(n) for storing the string and BITs

Key Concepts:
- Fenwick Tree (Binary Indexed Tree) for efficient prefix/suffix operations
- Difference arrays for range updates
- Palindrome check optimization using two difference arrays
*/

#include <cstring>
#include <iostream>
#include <string>
#include <utility>

using namespace std;
typedef long long LL;
typedef pair<int, int> PairII;
typedef pair<LL, LL> PairLL;
const int MAX_N = 2e5 + 3;
int n, q, op, l, r, x, d[MAX_N], e[MAX_N];
struct BIT {
  int tree[MAX_N];
  void update(int idx, int val) {
    for (; idx <= n; idx += idx & -idx) {
      tree[idx] += val;
    }
  }
  int query(int idx) {
    int result = 0;
    for (; idx; idx -= idx & -idx) {
      result += tree[idx];
    }
    return result;
  }
  int query(int left, int right) {
    if (left > right)
      return 0;
    return query(right) - query(left - 1);
  }
  void clear() { memset(tree, 0, sizeof(int) * (n + 1)); }
} bit1, bit2;
void update1(int idx, int val) {
  if (!d[idx])
    bit1.update(idx, -1);  // Remove the previous contribution if it was zero
  d[idx] = (d[idx] + val) % 26;
  if (!d[idx])
    bit1.update(idx, 1);   // Add the new contribution if it became zero
}
void update2(int idx, int val) {
  if (!e[idx])
    bit2.update(idx, -1);  // Remove the previous contribution if it was zero
  e[idx] = (e[idx] + val) % 26;
  if (!e[idx])
    bit2.update(idx, 1);   // Add the new contribution if it became zero
}
bool query1(int left, int right) { return !bit1.query(left + 1, right); }  // Check no adjacent same chars
bool query2(int left, int right) { return !bit2.query(left + 2, right); }  // Check no 3-char palindromes
string s;
void solve() {
  cin >> n >> q >> s;
  s = " " + s;  // Prepend space to make indexing easier
  for (int i = 2; i <= n; ++i) {
    d[i] = (s[i] - s[i - 1] + 26) % 26;  // Calculate difference with modulo 26
    if (!d[i])
      bit1.update(i, 1);  // Track zeros in d array (same adjacent chars)
  }
  for (int i = 3; i <= n; ++i) {
    e[i] = (s[i] - s[i - 2] + 26) % 26;  // Calculate difference with modulo 26
    if (!e[i])
      bit2.update(i, 1);  // Track zeros in e array (palindromic triplets)
  }
  while (q--) {
    cin >> op >> l >> r;
    if (op == 1) {
      cin >> x;
      x %= 26;  // Reduce x modulo 26 to avoid unnecessary operations
      if (l >= 2)
        update1(l, x);  // Update difference array for adjacent chars
      if (r + 1 <= n)
        update1(r + 1, 26 - x);  // Cancel out the impact on next element
      if (r - l >= 1) {  // At least one gap between l and r
        if (l >= 3)
          update2(l, x);  // Update difference array for palindromic triplets
        if (l + 1 >= 3)
          update2(l + 1, x);  // Update for l+1
        if (r + 1 <= n)
          update2(r + 1, 26 - x);  // Cancel out next element's impact
        if (r + 2 <= n)
          update2(r + 2, 26 - x);  // Cancel out element after next
      } else {  // l and r are adjacent
        if (l >= 3)
          update2(l, x);  // Update difference array
        if (r + 2 <= n)
          update2(r + 2, 26 - x);  // Cancel out next element
      }
    } else {
      if (query1(l, r) && query2(l, r))  // Check both conditions
        cout << "YES\n";
      else
        cout << "NO\n";
    }
  }
  bit1.clear();  // Clear the BITs for next test case
  bit2.clear();
}
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);
  int T;
  cin >> T;
  while (T--)
    solve();
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByRound/1881/1881_G_Anya_and_the_Mysterious_String.py
# Problem: CF 1881 G - Anya and the Mysterious String
# https://codeforces.com/contest/1881/problem/G

"""
Code Purpose:
This code solves the problem of determining whether a substring of a string is "beautiful"
(i.e., does not contain any palindromic substring of length at least 2) after applying
various operations on the string. It uses a segment tree with lazy propagation techniques,
and tracks differences between consecutive characters in the string to efficiently detect
palindromes.

Algorithms/Techniques:
- Binary Indexed Tree (Fenwick Tree) for efficient range updates and queries
- Difference array technique to track changes in character differences
- Efficient palindrome detection using difference tracking

Time Complexity: O((n + m) * log(n)) where n is the length of string and m is number of queries
Space Complexity: O(n) for storing the string, difference arrays and segment trees
"""

import io
import os
import sys

input = io.BytesIO(os.read(0, os.fstat(0).st_size)).readline


def get_sum0(i, tree):
    """Get prefix sum up to index i in the Fenwick tree."""
    s = 0
    while i > 0:
        s += tree[i]
        i -= i & -i
    return s


def get_sum(s, t, tree):
    """Get sum in range [s, t] using prefix sums."""
    ans = 0 if s > t else get_sum0(t, tree) - get_sum0(s - 1, tree)
    return ans


def add(i, x, tree):
    """Add value x to index i in the Fenwick tree."""
    while i < len(tree):
        tree[i] += x
        i += i & -i


t = int(input())
ans = []
for _ in range(t):
    n, m = map(int, input().split())
    s = list(input().rstrip())
    # Convert characters to 0-25 range for easier computation
    s = [0] + [i - 97 for i in s]
    # x[i] represents difference between s[i+1] and s[i]
    x = [-1] * (n + 1)
    for i in range(1, n):
        x[i] = (s[i + 1] - s[i]) % 26
    # y[i] represents difference between s[i+2] and s[i] (used for detecting palindromes of len 3+)
    y = [-1] * (n + 1)
    for i in range(1, n - 1):
        y[i] = (s[i + 2] - s[i]) % 26
    # Segment trees for tracking changes in difference arrays
    tree1, tree2 = [0] * (n + 5), [0] * (n + 5)
    # Initialize segment trees with initial values
    for i in range(1, n):
        if not x[i]:
            tree1[i] = 1
        if not y[i]:
            tree2[i] = 1
    # Build BIT from initial values
    for i in range(1, n + 5):
        j = i + (i & -i)
        if j < n + 5:
            tree1[j] += tree1[i]
            tree2[j] += tree2[i]
    for _ in range(m):
        t = list(map(int, input().split()))
        l, r = t[1], t[2]
        if t[0] == 1:
            u = t[3] % 26
            if not u:
                continue
            # Update x and y arrays for affected indices
            for i in [l - 1, r]:
                d = u if i < l else -u
                if not 1 <= i < n:
                    continue
                if not x[i]:
                    add(i, -1, tree1)  # Remove from tree if it was zero
                x[i] = (x[i] + d) % 26  # Update difference
                if not x[i]:
                    add(i, 1, tree1)  # Add to tree if it became zero
            z = [l - 2, l - 1, r - 1, r] if l ^ r else [l - 2, r]
            # Process y array changes
            for i in z:
                d = u if i < l else -u
                if not 1 <= i < n - 1:
                    continue
                if not y[i]:
                    add(i, -1, tree2)
                y[i] = (y[i] + d) % 26
                if not y[i]:
                    add(i, 1, tree2)
        else:
            # Query for beautiful substring
            # Count zeros in x and y arrays to detect palindromes
            c = get_sum(l, r - 1, tree1) + get_sum(l, r - 2, tree2)
            ans0 = "YES" if not c else "NO"
            ans.append(ans0)
sys.stdout.write("\n".join(ans))


# https://github.com/VaHiX/CodeForces/